File: lib/taskgroups/os.rb
module OSType
  OTHER = 0
  RHEL = 1
  GENTOO = 2
  UBUNTU = 3
end

class Os
  def self.l
    @l ||= SemanticLogger["os"]
  end

  def self.get_os
    case File.foreach("/etc/os-release").lazy.grep(/^ID=/).first
    when /fedora/
      OSType::RHEL
    when /gentoo/
      OSType::GENTOO
    when /rocky/
      OSType::RHEL
    when /ubuntu/
      OSType::UBUNTU
    else
      OSType::OTHER
    end
  end

  def self.pkg_update
    l.info "pulling from package mgr repos, then upgrading OS packages"
    case get_os
    when OSType::RHEL
      system "sudo dnf update -y"
    when OSType::GENTOO
      if Date.today > (DateTime.parse File.read "/var/db/repos/gentoo/metadata/timestamp.chk")
        system "sudo emaint --auto sync"
      end
      system "sudo emerge -vuDN @world"
    when OSType::UBUNTU
      system "sudo apt update"
      system "sudo apt upgrade -y"
    end
  end

  def self.pkg_clean
    case get_os
    when OSType::RHEL
      system "sudo dnf autoremove"
    when OSType::GENTOO
      system "sudo emerge --depclean"
    end
  end

  def self.all
    pkg_update
    pkg_clean
  end
end

File: src/task/os.rs
use crate::{errors::Error, helpers::exec};
use log::{info, warn};
use std::{
    fs::{self, File},
    io::{self, BufRead},
    process::Command,
};
use time::{format_description::well_known::Rfc2822, Duration, OffsetDateTime};

pub fn update() -> Result<(), Error> {
    info!("updating and cleaning OS packages");
    let os = get_os()?;
    match os {
        OS::RHELLike => {
            info!("detected RHEL-like OS");
            exec(Command::new("sudo").args(["dnf", "update", "-y"]))?;
            exec(Command::new("sudo").args(["dnf", "autoremove", "-y"]))?;
        }
        OS::Gentoo => {
            info!("detected Gentoo (you are hard-core)");
            let timestamp = fs::read_to_string("/var/db/repos/gentoo/metadata/timestamp.chk")?;
            let timestamp = OffsetDateTime::parse(&timestamp, &Rfc2822)?;
            let should_pull_packages =
                match OffsetDateTime::now_utc().checked_sub(Duration::days(1)) {
                    Some(one_day_ago) => timestamp < one_day_ago,
                    None => {
                        warn!("could not execute simple date math");
                        true
                    }
                };
            if should_pull_packages {
                exec(Command::new("sudo").args(["dnf", "autoremove", "-y"]))?;
            }
        }
        OS::Ubuntu => {
            info!("detected Ubuntu");
            exec(Command::new("sudo").args(["apt", "update"]))?;
            exec(Command::new("sudo").args(["apt", "upgrade", "-y"]))?;
            exec(Command::new("sudo").args(["apt", "autoremove", "-y"]))?;
        }
    }
    Ok(())
}

#[derive(Debug)]
enum OS {
    RHELLike,
    Gentoo,
    Ubuntu,
}

fn get_os() -> Result<OS, Error> {
    let file = File::open("/etc/os-release")?;
    let id = io::BufReader::new(file).lines().find(|line| match line {
        Ok(line) => line.starts_with("ID="),
        Err(_) => false,
    });
    match id {
        Some(id) => match id {
            Ok(id) => {
                if id.contains("rhel") || id.contains("rocky") || id.contains("fedora") {
                    Ok(OS::RHELLike)
                } else if id.contains("ubuntu") || id.contains("debian") {
                    Ok(OS::Ubuntu)
                } else if id.contains("gentoo") {
                    Ok(OS::Gentoo)
                } else {
                    Err(Error::UnsupportedOS)
                }
            }
            Err(e) => Err(e.into()),
        },
        None => Err(Error::CouldNotDetermineOS),
    }
}
